esp-hal-common/build.rs

```diff
@@ -12,40 +12,90 @@ fn main() {
     }
 
-    // Configuration symbol for the enabled chip
-    if esp32 {
-        println!("cargo:rustc-cfg=esp32");
+    // Define all required configuration symbols for the enabled chip.
+    //
+    // When adding a new device, at the bare minimum the following symbols MUST be
+    // defined:
+    //   - the name of the device
+    //   - the architecture ('riscv' or 'xtensa')
+    //   - the core count ('single_core' or 'multi_core')
+    //
+    // Additionally, the following symbols MAY be defined if present:
+    //   - 'dac'
+    //   - 'gdma'
+    //   - 'i2c1'
+    //   - 'pdma'
+    //   - 'rmt'
+    //   - 'spi3'
+    //   - 'systimer'
+    //   - 'timg1'
+    //   - 'uart2'
+    //   - 'usb_otg'
+    //   - 'usb_serial_jtag'
+    //
+    // New symbols can be added as needed, but please be sure to update both this
+    // comment and the required vectors below.
+    let symbols = if esp32 {
+        vec![
+            "esp32",
+            "xtensa",
+            "multi_core",
+            "dac",
+            "i2c1",
+            "pdma",
+            "rmt",
+            "spi3",
+            "timg1",
+            "uart2",
+        ]
     } else if esp32c2 {
-        println!("cargo:rustc-cfg=esp32c2");
+        vec!["esp32c2", "riscv", "single_core", "gdma", "systimer"]
     } else if esp32c3 {
-        println!("cargo:rustc-cfg=esp32c3");
+        vec![
+            "esp32c3",
+            "riscv",
+            "single_core",
+            "gdma",
+            "rmt",
+            "spi3",
+            "systimer",
+            "timg1",
+            "usb_serial_jtag",
+        ]
     } else if esp32s2 {
-        println!("cargo:rustc-cfg=esp32s2");
+        vec![
+            "esp32s2",
+            "xtensa",
+            "single_core",
+            "dac",
+            "i2c1",
+            "pdma",
+            "rmt",
+            "spi3",
+            "systimer",
+            "timg1",
+            "usb_otg",
+        ]
     } else if esp32s3 {
-        println!("cargo:rustc-cfg=esp32s3");
-    }
-
-    // Configuration symbol for the architecture of the enabled chip
-    if esp32c2 || esp32c3 {
-        println!("cargo:rustc-cfg=riscv");
-    } else {
-        println!("cargo:rustc-cfg=xtensa");
-    }
-
-    // Inject a configuration symbol to state the core count of the enabled chip
-    if esp32c2 || esp32c3 || esp32s2 {
-        println!("cargo:rustc-cfg=single_core");
+        vec![
+            "esp32s3",
+            "xtensa",
+            "multi_core",
+            "gdma",
+            "i2c1",
+            "rmt",
+            "spi3",
+            "systimer",
+            "timg1",
+            "uart2",
+            "usb_otg",
+            "usb_serial_jtag",
+        ]
     } else {
-        println!("cargo:rustc-cfg=multi_core");
-    }
-
-    // Configuration symbol for the SYSTIMER peripheral
-    if esp32c2 || esp32c3 || esp32s2 || esp32s3 {
-        println!("cargo:rustc-cfg=has_systimer");
-    }
+        vec![]
```

Comment:
maybe panic here since not defining anything for a chip is a major problem?

Reply:
It can't actually get that far if no chip is defined, we check that prior to this `if` block and it will panic [here](https://github.com/esp-rs/esp-hal/blob/b79becaae2f058a8a6c4512d40cfd03cb3ed304c/esp-hal-common/build.rs#L12) if no chip was specified. I guess `unreachable!()` would be appropriate.

Reply:
yes `unreachable!()` is even better